Order Creation: Align text fields in the address form

ealeksandrov opened the pull request:
Closes: https://github.com/woocommerce/woocommerce-ios/issues/6018.

## Description

This PR updates Address Form to align textfields vertically. This also fixes issue with missing trailing spaces input.

## Changes
- Adds `MaxWidthPreferenceKey` to a title view in `TitleAndTextFieldRow`.
- Updates `SingleAddressForm` to store shared (max) title width and inject it back into all fields.

### Known limitations
- `TitleAndValueRow` (Country and City fields) wasn't updated yet, but it doesn't have "missing spaces" bug and can be updated in a same way after testing and approval of this PR.

## Testing

1. Open Address Form from New Order or Order Details flows.
2. Enter some data in different fields and confirm that alignment is correct.
3. Update dynamic type size and recheck the layout.

## Screenshots

before|after
--|--
![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/163026314-5e7783ab-094a-4b5e-b720-ce342853645f.png)|![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/163026350-90a4c09c-83d5-4bfc-ad0b-adf9711531e4.png)
![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/163026442-073442a4-c6d4-4c56-9726-606d796677ee.png)|![Simulator Screen Shot](https://user-images.githubusercontent.com/3132438/163026467-5d32f40c-9bb5-4f61-9d96-98a9f5597488.png)


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ealeksandrov committed:
Apply max shared title width for fields in EditOrderAddressForm

ealeksandrov committed:
Store shared TitleAndTextFieldRow title width in MaxWidthPreference

ealeksandrov commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/TitleAndTextFieldRow.swift:
```diff
@@ -58,6 +64,33 @@ private extension TitleAndTextFieldRow {
+        GeometryReader { geometry in
+            Color.clear
+                .preference(key: MaxWidthPreferenceKey.self,
+                            value: geometry.size.width)
```
Not a fan of helper views, but wasn't able to find a better solution.
There are "Anchor preferences" that implicitly read view geometry, but they are more complicated and doesn't provide single "width" or "layout guide" value. They fit best to convert rect/point from one view context to another, but we don't need it here.

Ecarrion reviewed (approved):
Smart solution @ealeksandrov! It feels a little bit magical and somehow disconnected because `MaxWidthPreferenceKey` is read in one view and read from another in a bottom-up style but I don't have any better suggestion 😅 

I wonder if something similar could have been achieved by environment values? 🤔 

Aside from ☝️, I noticed a couple of small UI Issues.

1. If we use **Dynamic Type XXL** the labels overlap a bit

<img width="435" alt="Screen Shot 2022-04-13 at 10 16 45 AM" src="https://user-images.githubusercontent.com/562080/163215646-14410660-9546-42b1-b772-645fcd0ec164.png">

2. It looks a bit odd w
<img width="405" alt="Screen Shot 2022-04-13 at 10 17 38 AM" src="https://user-images.githubusercontent.com/562080/163215914-66229368-21b9-4951-83be-764f664e837a.png">
hen the country has no states to select.

ealeksandrov committed:
Apply MaxWidthModifier on field title before setting static width

ealeksandrov commented:
Thanks for the review @Ecarrion!

Agree about magical feeling 🙂 From what I understand - environment values are about passing data from top to bottom layers. Here we need to report data from bottom views to the parent.

1. The reason for this issue was just a wrong order of 2 lines 😄
Dynamic type sizing was already handled, it was a mistake preventing live update - dynamic width getter was applied *after* static width setter. Fixed in f37d7b4.

regular size|Dynamic Type XXL
--|--
![Simulator Screen Shot - 1](https://user-images.githubusercontent.com/3132438/163242802-fcb5bad4-ad1e-40c5-9528-b46039debefb.png)|![Simulator Screen Shot - 2](https://user-images.githubusercontent.com/3132438/163242812-4dc6eefc-9f51-4ed9-9042-e1d375c19035.png)

2. I plan to update layout for `TitleAndValueRow` separately, after we're happy with current implementation. It should be straightforward to add the same preference, modifier and binding.

Ecarrion commented:
> Agree about magical feeling

Maybe it will feel less magical if we create a private custom row inside the form where it's clearly visible (because it is in the same file) where the preference width key is being set.

What do you think?

ealeksandrov committed:
Move MaxWidthModifier to separate file for reusing

ealeksandrov committed:
Add spacing in TitleAndValueRow (same as in TitleAndTextFieldRow)

ealeksandrov committed:
Add MaxWidthModifier to TitleAndValueRow

ealeksandrov committed:
Add comments

ealeksandrov commented:
@Ecarrion but that will prevent reuse of this preference in other screens.
Also we need same update to `TitleAndValueRow` and duplicating second field will clutter single file a lot.

I've applied same change for `TitleAndValueRow`, moved preference into a separate file and added some comments ([commits subset](https://github.com/woocommerce/woocommerce-ios/pull/6625/files/000016528d6631d0eba41e3c5d56b5cac31eb330..889967ef91ec20e0c290370a498ad6fd677cb772)).
Do you think it's better or we can add more structure and/or clarifications?

regular size|Dynamic Type XXL
--|--
![Simulator Screen Shot - 1](https://user-images.githubusercontent.com/3132438/163395592-77a6aab6-0181-4133-8340-76edf959a187.png)|![Simulator Screen Shot - 2](https://user-images.githubusercontent.com/3132438/163395610-c0c35a0c-f06e-40c1-8b8c-51963f7cf898.png)

ealeksandrov committed:
Merge branch 'trunk' into issue/6018-fields-alignment-in-address-form

ealeksandrov committed:
Fix typo
